packages/rich-text-structure/src/create.js

```diff
@@ -0,0 +1,326 @@
+			) {
+				accumulator.end = accumulator.text.length;
+			}
+		}
+
+		if ( node.nodeName === 'BR' ) {
+			if ( unwrapNodeMatch( node ) ) {
+				return accumulator;
+			}
+
+			accumulator.text += '\n';
+			accumulator.formats.length += 1;
+			return accumulator;
+		}
+
+		let format;
+
+		if ( ! unwrapNodeMatch( node ) ) {
+			const type = node.nodeName.toLowerCase();
+			const attributes = getAttributes( node, { removeAttributeMatch } );
+			format = attributes ? { type, attributes } : { type };
+		}
+
+		const record = createRecord( node, range, settings );
+		const text = record.text;
+		const start = accumulator.text.length;
+
+		// Expand range if it ends in this node.
+		if ( range ) {
+			if (
+				node.parentNode === range.endContainer &&
+				node === range.endContainer.childNodes[ range.endOffset - 1 ]
+			) {
+				accumulator.end = start + text.length;
+			}
+		}
+
+		// Don't apply the element as formatting if it has no content.
+		if ( isEmpty( record ) && format && ! format.attributes ) {
+			return accumulator;
+		}
+
+		const { formats } = accumulator;
+
+		if ( format && format.attributes && text.length === 0 ) {
+			format.object = true;
+
+			if ( formats[ start ] ) {
+				formats[ start ].unshift( format );
+			} else {
+				formats[ start ] = [ format ];
+			}
+		} else {
+			accumulator.text += text;
+
+			let i = record.formats.length;
+
+			while ( i-- ) {
+				const index = start + i;
+
+				if ( format ) {
+					if ( formats[ index ] ) {
+						formats[ index ].push( format );
+					} else {
+						formats[ index ] = [ format ];
+					}
+				}
+
+				if ( record.formats[ i ] ) {
+					if ( formats[ index ] ) {
+						formats[ index ].push( ...record.formats[ i ] );
+					} else {
+						formats[ index ] = record.formats[ i ];
+					}
+				}
+			}
+		}
+
+		if ( record.start !== undefined ) {
+			accumulator.start = start + record.start;
+		}
+
+		if ( record.end !== undefined ) {
+			accumulator.end = start + record.end;
+		}
+
+		return accumulator;
+	}, emptyRecord );
+}
+
+/**
+ * Gets the attributes of an element in object shape.
+ *
+ * @param {HTMLElement} element                       Element to get attributes from.
+ * @param {Function}    settings.removeAttributeMatch Function whose return value
+ *                                                    determines whether or not to
+ *                                                    remove an attribute based on name.
+ *
+ * @return {?Object} Attribute object or `undefined` if the element has no
+ *                   attributes.
+ */
+function getAttributes( element, {
+	removeAttributeMatch = () => false,
+} ) {
+	if ( ! element.hasAttributes() ) {
+		return;
+	}
+
+	return Array.from( element.attributes ).reduce( ( accumulator, { name, value } ) => {
+		if ( ! removeAttributeMatch( name ) ) {
+			accumulator = accumulator || {};
+			accumulator[ name ] = value;
+		}
+
+		return accumulator;
+	}, undefined );
```

Comment:
for fun…

```js
const add = ( attributes, { name, value } ) =>
	! removeAttributeMatch( name )
		? { ...( attributes || {} ), name: value }
		: attributes;
```

or for speed…

```js
const { attributes } = element;
const validAttributes = {}
let hasValidAttributes = false;
for ( let i = 0; i < length; i++ ) {
	const { name, value } = attributes[ i ];

	if ( removeAttributeMatch( name ) ) {
		continue;
	}

	validAttributes[ name ] = value;
	hasValidAttributes = true;
}

return hasValidAttributes
	? validAttributes
	: undefined
```

Reply:
What's best for readability? :)

Reply:
oh if I could only give you a black and white answer! :wink:

this is style I'm sure. personally I get really confused trying to understand the intention behind code I see in 99% of the `for` loops I read. I don't mind a certain amount of mutation but find that when there is mutation separated by more than a couple lines then it gets much harder for me to understand what's happening (i.e. what could go wrong)

my philosophy is that it's the computer's job to interpret the code, not mine, so I don't want to have to build a running virtual machine in my brain while reading code, which is why I prefer single-static assignment and immutability and higher-order concepts like `map` and `reduce` and `filter` to imperative loop structures.

in this example I'd say both are understandable. the `add` function above describes each step in the iterative process so we can try to understand the whole by understanding the part. in the imperative version for example we have to track the control flow of the `continue` statement.

the `for` loop's performance characteristics are more predictable because we can get a reasonable idea of how the JS engine will compile and execute this code. it involves more accounting because we've encoded more of the task itself into the code whereas above we've encoded more of the intended output instead. the top snippet trades performance for what should be a smaller surface area for introducing bugs.

🤷‍♂️ 😄